File: src/model/transaction_status.py
"""
Transaction Status Enum - Status transaksi dalam sistem
"""

from enum import Enum


class TransactionStatus(Enum):
    """
    Enum untuk status transaksi
    """
    
    PENDING = "PENDING"
    """Transaksi menunggu untuk diproses"""
    
    PROCESSING = "PROCESSING"
    """Transaksi sedang diproses"""
    
    PREPARED = "PREPARED"
    """Transaksi sudah diprepare (2PC phase 1 selesai)"""
    
    COMMITTED = "COMMITTED"
    """Transaksi berhasil di-commit"""
    
    ABORTED = "ABORTED"
    """Transaksi di-abort"""
    
    FAILED = "FAILED"
    """Transaksi gagal"""
    
    TIMEOUT = "TIMEOUT"
    """Transaksi timeout"""
    
    def is_active(self) -> bool:
        """Cek apakah transaksi masih aktif"""
        return self in [TransactionStatus.PENDING, 
                       TransactionStatus.PROCESSING,
                       TransactionStatus.PREPARED]
    
    def is_final(self) -> bool:
        """Cek apakah status adalah final state"""
        return self in [TransactionStatus.COMMITTED,
                       TransactionStatus.ABORTED,
                       TransactionStatus.FAILED,
                       TransactionStatus.TIMEOUT]

File: src/main/java/com/simulator/model/TransactionStatus.java
package com.simulator.model;

/**
 * Enum untuk status transaksi
 */
public enum TransactionStatus {
    INITIALIZED("Initialized"),
    PREPARING("Preparing"),
    PREPARED("Prepared"),
    COMMITTING("Committing"),
    COMMITTED("Committed"),
    ABORTING("Aborting"),
    ABORTED("Aborted"),
    TIMEOUT("Timeout"),
    FAILED("Failed");
    
    private final String description;
    
    TransactionStatus(String description) {
        this.description = description;
    }
    
    public String getDescription() {
        return description;
    }
    
    public boolean isCompleted() {
        return this == COMMITTED || this == ABORTED || this == FAILED;
    }
    
    public boolean isFinal() {
        return isCompleted() || this == TIMEOUT;
    }
    
    @Override
    public String toString() {
        return description;
    }
}